pennylane/measurements/sample.py

```diff
@@ -106,4 +111,47 @@ def circuit(x):
             )
         wires = Wires(wires)
 
-    return MeasurementProcess(Sample, obs=op, wires=wires)
+    return _Sample(Sample, obs=op, wires=wires)
+
+
+# TODO: Make public when removing the ObservableReturnTypes enum
+class _Sample(MeasurementProcess):
+    """Measurement process that returns the samples of a given observable."""
+
+    def process(self, samples: np.ndarray, shot_range: Tuple[int] = None, bin_size: int = None):
+        name = self.obs.name if self.obs is not None else None
+        # Select the samples from samples that correspond to ``shot_range`` if provided
+        if shot_range is not None:
+            # Indexing corresponds to: (potential broadcasting, shots, wires). Note that the last
+            # colon (:) is required because shots is the second-to-last axis and the
+            # Ellipsis (...) otherwise would take up broadcasting and shots axes.
+            samples = samples[..., slice(*shot_range), :]
+
+        if len(self.wires) != 0:
+            # if wires are provided, then we only return samples from those wires
+            samples = samples[..., np.array(self.wires)]
+
+        num_wires = samples.shape[-1]  # wires is the last dimension
+
+        if self.obs is None:
+            # if no observable was provided then return the raw samples
+            return samples if bin_size is None else samples.reshape(num_wires, bin_size, -1)
```

Comment:
Right now we can have all the following `sample.shape`:
```python
if `self.obs` is None and `bin_size` is None:
    shape = samples.shape
elif `self.obs` is None and `bin_size` is not None:
    shape = (num_wires, bin_size, -1)
elif `self.obs` is not None and `bin_size` is None:
    shape = samples.shape
elif `self.obs` is not None and `bin_size` is not None:
    shape = (bin_size, -1)
```

Reply:
I don't know, but reshaping the `samples` array which has shape `(shots, num_wires)` to an array with shape `(num_wires, bin_size, num_bins)` doesn't even make sense in the first place. There should be a transpose followed by a reshape.

Maybe the original author meant to put the `num_wires` axis last? I'm not really sure.

Reply:
Don't know, but if we change it we should change all the logic that use the results of the `QubitDevice.sample` method. I would leave it like this for now.